libgnucash/app-utils/gnc-option.hpp

```diff
@@ -0,0 +1,254 @@
+                               int>  = 0>
+
+    GncOption(OptionType option) :
+        m_option{std::make_unique<GncOptionVariant>(option)} {}
+    template <typename ValueType,
+              typename std::enable_if_t<!is_OptionClassifier_v<ValueType>,
+                               int>  = 0>
+    GncOption(const char* section, const char* name,
+              const char* key, const char* doc_string,
+              ValueType value,
+              GncOptionUIType ui_type = GncOptionUIType::INTERNAL);
+    template <typename ValueType> void set_value(ValueType value);
+    template <typename ValueType> void set_default_value(ValueType value);
+    template <typename ValueType> ValueType get_default_value() const;
+    template <typename ValueType> ValueType get_value() const;
+    void reset_default_value();
+
+    const std::string& get_section() const;
+    const std::string& get_name() const;
+    const std::string& get_key() const;
+    const std::string& get_docstring() const;
+    void set_ui_item(GncOptionUIItemPtr&& ui_elem);
+    const GncOptionUIType get_ui_type() const;
+    void set_ui_item_selectable(bool) const noexcept;
+    GncOptionUIItem* const get_ui_item() const;
+    void set_ui_item_from_option();
+    void set_option_from_ui_item();
+    void make_internal();
+    bool is_changed() const noexcept;
+/** @returns false unless m_option contains a GncOptionMultiselectValue or
+ * GncOptionAccountListValue for which multiple selections have been enabled.
+ */
+    bool is_multiselect() const noexcept;
+/** Implemented only for GncOptionNumericRange */
+    template <typename ValueType> void get_limits(ValueType&, ValueType&,
+                                                  ValueType&) const noexcept;
+/** Not implemented for GncOptionValue. */
+    template <typename ValueType> bool validate(ValueType value) const;
+/** Implemented only for GncOptionMultiselectValue. */
+    std::size_t num_permissible_values() const;
+/** Implemented only for GncOptionMultiselectValue. */
+    std::size_t permissible_value_index(const char* value) const;
+/** Implemented only for GncOptionMultiselectValue. */
+    const char* permissible_value(std::size_t index) const;
+/** Implemented only for GncOptionMultiselectValue. */
+    const char* permissible_value_name(std::size_t index) const;
+/** Implemented only for GncOptionAccountListValue. */
+    GList* account_type_list() const noexcept;
+    bool is_alternate() const noexcept;
+    void set_alternate(bool) noexcept;
+    std::ostream& out_stream(std::ostream& oss) const;
+    std::istream& in_stream(std::istream& iss);
+
+    friend GncOptionVariant& swig_get_option(GncOption*);
+
+private:
+    inline static const std::string c_empty_string{""};
+    GncOptionVariantPtr m_option;
+    GncOptionUIItemPtr m_ui_item{nullptr};
+};
+
+inline bool
+operator<(const GncOption& right, const GncOption& left)
+{
+    return right.get_key() < left.get_key();
```

Comment:
What happens if sort keys are equal ? Should we sort on name secondly ?

Reply:
What happens is that `operator<` returns `false` and `left` isn't moved before `right` so the options are displayed on the dialog tab in the order that they were registered.

It's a programming error to specify two options in the same sectopm with the same key so I guess we should crash instead. ;-)

Reply:
Right, my reasoning is flawed...